jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletChannelState.java

```diff
@@ -1369,6 +1336,8 @@ public void onReadIdle()
             if (LOG.isDebugEnabled())
                 LOG.debug("onReadIdle {}", toStringLocked());
 
+            // Blocking moves from UNREADY -> IDLE
```

Comment:
I don't think we should only move from UNREADY TO IDLE if the demand callback is called or fatal error.  Moving from READY to IDLE is OK, as you have to be READY to be able to consume any input.

Note the method name is not really good....

actually, see my other comments, as I don't really see why we need tristate.  Don't we just need to know if we are demanding or not?

Reply:
We do need 3 states, hopefully the documentation of the FSM I added should clear this up.